### PR TITLE
Add parsing of Polygon problems

### PR DESCRIPTION
**Pull request by BabichMikhail**

### CODE REVIEW

File: CATS/DB.pm [hunk 1]
```diff
@@ -40,29 +40,8 @@ sub object_by_id
...
-    elsif ($CATS::Config::db_dsn =~ /Oracle/)
-    {
-        $dbh->selectrow_array(q~SELECT key_seq.nextval FROM DUAL~);
-    }
-    else
-    {
-        die 'Error in new_id';
-    }
+sub new_id {
+    Digest::MD5::md5_hex(Encode::encode_utf8($_[1]));
```
Comment (klenin) [inline CATS/DB.pm]: This will totally break cats-main. Id generation must still work with the database.

File: CATS/Problem/PolygonParser.pm [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+sub tag_handlers() {{
+    problem => { s => \&start_tag_problem },
+    names => {},
+    name => { s => \&start_tag_name },
+    statements => {},
+    statement => {},
+    judging => { s => \&start_tag_judging },
+    testset => { s => \&start_tag_testset, e => \&end_tag_testset },
```
Comment (klenin) [inline CATS/Problem/PolygonParser.pm]: `judging/testset` ?

File: CATS/Problem/PolygonParser.pm [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+sub start_tag_solution_source {
+    (my CATS::Problem::PolygonParser $self, my $atts) = @_;
+    $atts->{src} = $atts->{path};
+    ($atts->{name}) = $atts->{path} =~ m/\/(.*)\..*$/;
+    my $sol = $self->set_named_object($atts->{name}, {
+        $self->problem_source_common_params($atts, 'solution'),
+    });
+    push @{$self->{problem}{solutions}}, $sol;
+    ${$self->{tag_stack}}[-2]->{atts}{tag} eq 'main' or return;
```
Comment (klenin) [inline CATS/Problem/PolygonParser.pm:228]: Why this check is so late?

File: CATS/Problem/PolygonParser.pm [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    $atts->{path} and $atts->{type} or return;
+    $atts->{path} =~ m/testlib.h/ or return;
+    $self->{testlib} = $atts->{path};
+    $self->{has_testlib} = 1;
+}
+
+sub start_tag_checker_source {
+    (my CATS::Problem::PolygonParser $self, my $atts) = @_;
+    $atts->{src} = $atts->{path};
+    ($atts->{name}) = $atts->{path} =~ m/\/(.*)\..*$/;
```
Comment (klenin) [inline CATS/Problem/PolygonParser.pm:208]: Should probably extract this code.

File: CATS/Problem/PolygonParser.pm [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+    });
+    $generator->{name} = $atts->{name};
+    push @{$self->{problem}{generators}}, $generator;
+}
+
+sub on_start_tag {
+    my CATS::Problem::PolygonParser $self = shift;
+    my ($p, $el, %atts) = @_;
+    my $h = tag_handlers()->{$el} ? tag_handlers()->{$el} :
+        tag_handlers()->{ $el = "${$self->{tag_stack}}[-1]->{name}/$el" }
```
Comment (klenin) [inline CATS/Problem/PolygonParser.pm]: Why do you need to change `$el` ?

File: CATS/Problem/PolygonParser.pm [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+sub on_end_tag {
+    my CATS::Problem::PolygonParser $self = shift;
+    my ($p, $el, %atts) = @_;
+    my $h = tag_handlers()->{$el} ? tag_handlers()->{$el} :
+        tag_handlers()->{ $el = "${$self->{tag_stack}}[-2]->{name}/$el" }
+        or $self->error("Unknown tag $el");
+    $h->{e} and $h->{e}->($self, \%atts, $el);
+    my $elem = ${$self->{tag_stack}}[-1];
```
Comment (klenin) [inline CATS/Problem/PolygonParser.pm]: This line does nothing?

File: CATS/Problem/PolygonParser.pm [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+    my $xml_parser = XML::Parser::Expat->new;
+    $xml_parser->setHandlers(
+        Start => sub { $self->on_start_tag(@_) },
+        End => sub { $self->on_end_tag(@_) },
+        Char => sub { $self->{Char} = $_[1]; },
+    );
+    $xml_parser->parse($self->{source}->read_member($xml_file));
+
+    if ($self->{has_testlib}) {
+        my @module_types = ('checker', 'validator', 'generator');
```
Comment (klenin) [inline CATS/Problem/PolygonParser.pm]: Use `qw`

File: CATS/Problem/PolygonParser.pm [hunk 8]
```diff
@@ -0,0 +1,328 @@
+package CATS::Problem::PolygonParser;
+
+use strict;
+use warnings;
+
+use File::Glob 'bsd_glob';
+
```
Comment (klenin) [inline CATS/Problem/PolygonParser.pm]: Is this needed?

File: CATS/Problem/PolygonParser.pm [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+use XML::Parser::Expat;
+
+use List::Util 'max';
```
Comment (klenin) [inline CATS/Problem/PolygonParser.pm]: Is this needed?

File: CATS/Problem/PolygonParser.pm (hunk 3 above)
Comment (BabichMikhail) [inline CATS/Problem/PolygonParser.pm:228]: Save all solutions but set only main solution for all tests

**Pull request merged**